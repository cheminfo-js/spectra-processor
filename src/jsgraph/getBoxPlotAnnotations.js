import { getBoxPlotData } from '../spectra/getBoxPlotData';

export function getBoxPlotAnnotations(spectra, options = {}) {
  const {
    q13FillColor = '#000',
    q13FillOpacity = 0.3,
    q2StrokeColor = '#000',
    q2StrokeWidth = 3,
    minMaxFillColor = '#000',
    minMaxFillOpacity = 0.15
  } = options;
  let annotations = [];

  let data = getBoxPlotData(spectra);
  if (q13FillOpacity && q13FillColor) {
    let q13 = [];
    for (let i = 0; i < data.x.length; i++) {
      q13.push({
        x: data.x[i],
        y: data.Q1[i]
      });
    }
    for (let i = data.x.length - 1; i >= 0; i--) {
      q13.push({
        x: data.x[i],
        y: data.Q3[i]
      });
    }
    annotations.push({
      type: 'polygon',
      layer: 0,
      properties: {
        fillColor: q13FillColor,
        fillOpacity: q13FillOpacity,
        strokeWidth: 0.0000001,
        position: q13
      }
    });
  }

  if (q2StrokeColor && q2StrokeWidth) {
    let q2 = [];
    for (let i = 0; i < data.x.length; i++) {
      q2.push({
        x: data.x[i],
        y: data.Q2[i]
      });
    }
    // Temp code because polyline is not working
    for (let i = data.x.length - 1; i >= 0; i--) {
      q2.push({
        x: data.x[i],
        y: data.Q2[i]
      });
    }

    annotations.push({
      type: 'polygon',
      layer: 0,
      properties: {
        strokeWidth: q2StrokeWidth,
        strokeColor: q2StrokeColor,
        position: q2
      }
    });
  }

  if (minMaxFillColor && minMaxFillOpacity) {
    let minMax = [];
    for (let i = 0; i < data.x.length; i++) {
      minMax.push({
        x: data.x[i],
        y: data.min[i]
      });
    }
    for (let i = data.x.length - 1; i >= 0; i--) {
      minMax.push({
        x: data.x[i],
        y: data.max[i]
      });
    }

    annotations.push({
      type: 'polygon',
      layer: 0,
      properties: {
        fillColor: minMaxFillColor,
        fillOpacity: minMaxFillOpacity,
        strokeWidth: 0.0000001,
        strokeColor: '#FFF',
        position: minMax
      }
    });
  }

  return annotations;
}
