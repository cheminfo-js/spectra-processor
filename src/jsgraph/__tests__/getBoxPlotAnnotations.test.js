import { getBoxPlotAnnotations } from '../getBoxPlotAnnotations';

let spectra = [
  {
    normalized: {
      x: [10, 20, 30],
      y: [1, 2, 3],
    },
  },
  {
    normalized: {
      x: [10, 20, 30],
      y: [2, 3, 4],
    },
  },
  {
    normalized: {
      x: [10, 20, 30],
      y: [3, 4, 5],
    },
  },
  {
    normalized: {
      x: [10, 20, 30],
      y: [4, 5, 6],
    },
  },
  {
    normalized: {
      x: [10, 20, 30],
      y: [6, 7, 8],
    },
  },
];

describe('getBoxPlotAnnotations', () => {
  it('check jsgraph annotations', () => {
    let result = getBoxPlotAnnotations(spectra);

    expect(result).toMatchSnapshot();
  });
});
