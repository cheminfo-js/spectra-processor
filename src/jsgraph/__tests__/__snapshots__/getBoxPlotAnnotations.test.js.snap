// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBoxPlotAnnotations check jsgraph annotations 1`] = `
Array [
  Object {
    "layer": 0,
    "properties": Object {
      "fillColor": "#000",
      "fillOpacity": 0.3,
      "position": Array [
        Object {
          "x": 10,
          "y": 1.5,
        },
        Object {
          "x": 20,
          "y": 2.5,
        },
        Object {
          "x": 30,
          "y": 3.5,
        },
        Object {
          "x": 30,
          "y": 7,
        },
        Object {
          "x": 20,
          "y": 6,
        },
        Object {
          "x": 10,
          "y": 5,
        },
      ],
      "strokeWidth": 1e-7,
    },
    "type": "polygon",
  },
  Object {
    "layer": 0,
    "properties": Object {
      "position": Array [
        Object {
          "x": 10,
          "y": 3,
        },
        Object {
          "x": 20,
          "y": 4,
        },
        Object {
          "x": 30,
          "y": 5,
        },
        Object {
          "x": 30,
          "y": 5,
        },
        Object {
          "x": 20,
          "y": 4,
        },
        Object {
          "x": 10,
          "y": 3,
        },
      ],
      "strokeColor": "#000",
      "strokeWidth": 3,
    },
    "type": "polygon",
  },
  Object {
    "layer": 0,
    "properties": Object {
      "fillColor": "#000",
      "fillOpacity": 0.15,
      "position": Array [
        Object {
          "x": 10,
          "y": 1,
        },
        Object {
          "x": 20,
          "y": 2,
        },
        Object {
          "x": 30,
          "y": 3,
        },
        Object {
          "x": 30,
          "y": 8,
        },
        Object {
          "x": 20,
          "y": 7,
        },
        Object {
          "x": 10,
          "y": 6,
        },
      ],
      "strokeColor": "#FFF",
      "strokeWidth": 1e-7,
    },
    "type": "polygon",
  },
]
`;
